Example for route mirroring policy (#23336)
envoyproxy/envoy: docs/root/start/sandboxes/route-mirror.rst

.. _install_sandboxes_route_mirror:

Route mirroring policies
========================

.. sidebar:: Requirements

   .. include:: _include/docker-env-setup-link.rst

This simple example demonstrates Envoy's request mirroring capability using
`request mirror policies <https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-routeaction-requestmirrorpolicy>`__.

Incoming requests are received by ``envoy-front-proxy`` service. The envoy instance running
in this container is configured with two routes:

.. literalinclude:: _include/route-mirror/front-envoy.yaml
   :language: yaml

>    :lines: 7-37

Reviewer: lets reduce the displayed lines - i would probably start on the `virtual_hosts` line - but i think at most 2-3 lines before and 1 after the highlighed text

lets also add a caption with download (you should be able to find existing examples)

> .. literalinclude:: _include/route-mirror/front-envoy.yaml

Reviewer: im not sure about putting the example code here - i think it distracts a bit from the explanation (and doesnt work so well with formatting in at least my browser)

im wondering whether we should keep it and/or whether we might want to put it in the corresponding steps - ie show the static config in the static config step, etc

Author: I took a stab at inlining it. for example:

![image](https://user-images.githubusercontent.com/76932416/197695720-2178b2a1-f28a-40ac-b425-c1ea62eba78a.png)

   :linenos:
   :emphasize-lines: 5-17
   :caption: Envoy configuration with route mirror policy

> A request for the path ``/service/1`` is forwarded to the ``service1`` cluster.
> In addition, the request is also forwarded to the ``service1-mirror`` cluster.

Reviewer: ```suggestion
Requests for the path ``/service/1`` are statically mirrored.

Each request handled by the ``service1`` cluster, and in addition, forwarded to the ``service1-mirror`` cluster.
```

> A request for the path ``/service/2`` is forwarded to the ``service2`` cluster.
> If a header ``x-mirror-cluster`` is specified in the request, envoy extracts the
> header value and forwards the request to a cluster with the same name, if found.
> For example, if we send a header with a request as ``x-mirror-cluster: service2-mirror``,
> the request will be forwarded to the ``service2-mirror`` cluster.

Reviewer: ```suggestion
Requests for the path ``/service/2`` are dynamically mirrored according to the presence and value of the ``x-mirror-cluster`` header.

All reqests for this path are  forwarded to the ``service2`` cluster, and are also mirrored to the cluster named in the header.

For example, if we send a header with a request as ``x-mirror-cluster: service2-mirror``,
the request will be forwarded to the ``service2-mirror`` cluster.
```

Im wondering if we need to say something about this pattern only being useful in a trusted environment - im not sure you would want an end-user determining where a request was mirrored, but i can see how it might be useful for a proxy to add this header for upstream processing, for example.

Author: Good point to mention. I will take a stab at it.

Step 1: Build the sandbox
*************************

Change to the ``examples/route-mirror`` directory.

.. code-block:: console

    $ pwd
    envoy/examples/route-mirror
    $ docker-compose build
    $ docker-compose up -d

> 

Reviewer: unnecessary whitespace

.. code-block:: console

> 

Reviewer: unnecessary whitespace

    $ pwd
    envoy/examples/route-mirror
    $ docker-compose ps

> 

Reviewer: unnecessary whitespace

>     envoy/examples/route-mirror

Reviewer: i think you can just remove these lines

    NAME                                COMMAND                  SERVICE             STATUS              PORTS
    ---------------------------------------------------------------------------------------------------------------------------

> 

Reviewer: unnecessary whitespace

    route-mirror-front-envoy-1       "/docker-entrypoint.…"   front-envoy         running             0.0.0.0:8001->8001/tcp,

>                                                                                                          :::8001->8001/tcp,

Reviewer: i think the ports are now incorrect here - also i think dont wrap it

Step 2: Make a request to the statically mirrored route
*******************************************************

Send a request to ``service1``:

.. code-block:: console

  $ pwd

>   envoy/examples/route-mirror

Reviewer: you can remove these lines

Author: I have removed them from the `curl` commands, but kept them for the `docker-compose` commands, as they have to be run from the specific directory.

Reviewer: we only need this once at the beginning of the sandbox - we can assume that the user is in the correct dir from there on - its kinda pseudo code anyway

  $ curl localhost:10000/service/1
  Hello from behind Envoy (service 1)!

> The command above sends a request to the ``envoy-front-proxy`` service which forwards the request to

Reviewer: lets move this above the code block - i think it works better as an intro

``service1`` and also sends the request to the service 1 mirror, ``service1-mirror``.

Step 3: View logs for the statically mirrored request
*****************************************************

View the logs for the ``service1`` and ``service1-mirror`` services:

.. code-block:: console

   $ docker-compose logs service1

> 

Reviewer: i think remove this whitespace

> Step 2: Demonstrate static mirror cluster name

Reviewer: ```suggestion
Step 2: Make a request to the statically mirrored route
```

> 

Reviewer: i think this needs some preamble

   ...
   Host: localhost:10000
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/1 HTTP/1.1" 200 -

   $ docker-compose logs service1-mirror

   ...
   Host: localhost-shadow:10000
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/1 HTTP/1.1" 200 -

> The above logs from the ``service1`` and ``service1-mirror`` services show that

Reviewer: same with these paragraphs - i think give the explanation first

> Step 3: See Logs

Reviewer: ```suggestion
Step 3: View logs for the statically mirrored request
```

> 

Reviewer: preamble

   ...
   Host: localhost-shadow:8080
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/1 HTTP/1.1" 200 -

The above logs from the ``service1`` and ``service1-mirror`` services show that
both the ``service1`` and ``service1-mirror`` services got the request.

You can also see that for the request to the ``service1-mirror`` service, the
``Host`` header was modified by Envoy to have a ``-shadow`` suffix in the
hostname.

> Step 4: Demonstrate mirror cluster via header

Reviewer: ```suggestion
Step 4: Make a request to the route mirrored by request header
```

*********************************************

In this step, we will see a demonstration where the request specifies via a header, ``x-mirror-cluster``,
the cluster that envoy will mirror the request to.

.. code-block:: console

  $ pwd
  envoy/examples/route-mirror
  $ curl --header "x-mirror-cluster: service2-mirror" localhost:8080/service/2
  Hello from behind Envoy (service 2)!

The command above sends a request to the ``envoy-front-proxy`` service which forwards the request to
``service2`` and also mirrors the request to the cluster named, ``service2-mirror``.

> Step 4: See Logs

Reviewer: ```suggestion
Step 5: View logs for the request mirrored by request header
```

****************

> 

Reviewer: preamble

.. code-block:: console

   $ docker-compose logs service2

   ...
   Host: localhost:8080
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/2 HTTP/1.1" 200 -

   $ docker-compose logs service2-mirror

   ...
   Host: localhost-shadow:8080
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/2 HTTP/1.1" 200 -

The above logs show that both the ``service2`` and ``service2-mirror`` services
got the request.

You can also see that for the request to the ``service2-mirror`` service, the
``Host`` header was modified by Envoy to have a ``-shadow`` suffix in the
hostname.

If you do not specify the ``x-mirror-cluster`` in the request to ``service2``,
``service2-mirror`` will not receive the request.

> 

Reviewer: could you add a `seealso` section with links to any other reference docs and api please

   :language: yaml
   :lines: 16-34
   :linenos:
   :emphasize-lines: 6-11
   :caption: Envoy configuration with static route mirror policy :download:`front-envoy.yaml <_include/route-mirror/front-envoy.yaml>`

Requests for the path ``/service/2`` are dynamically mirrored according to the presence and value of
the ``x-mirror-cluster`` header.

All reqests for this path are forwarded to the ``service2`` cluster, and are also mirrored
to the cluster named in the header.

For example, if we send a request with the header ``x-mirror-cluster: service2-mirror``,
the request will be forwarded to the ``service2-mirror`` cluster.

.. literalinclude:: _include/route-mirror/front-envoy.yaml
   :language: yaml
   :lines: 16-34
   :linenos:
   :emphasize-lines: 12-17
   :caption: Envoy configuration with header based route mirror policy :download:`front-envoy.yaml <_include/route-mirror/front-envoy.yaml>`

The pattern to allow a request header to determine the cluster it is mirrored to
is most useful in a trusted environment. For example, a higher level Envoy instance (or
another application acting as a proxy) might automatically add this header to requests for upstream

> processing by another Envoy instance configured with request mirror policies.

Author: > thinking more about this - there may other situations where you might allow dynamic mirroring according to end users' requests

yeah, i share your ambiguity here. May be it's better to have this warning, than not. I think.

.. note::

   Envoy will only return the response it receives from the primary cluster to the client.

   For this example, responses from ``service1`` and ``service2`` clusters will be sent
   to the client. A response returned by the ``service1-mirror`` or the ``service2-mirror``
   cluster is not sent back to the client.

>    This also means that any problems in request processing in the mirror cluster doesn't affect the

Reviewer: i updated my suggestion here to include a point about latency

the cluster that envoy will mirror the request to.

Let's send a request to the ``envoy-front-proxy`` service which forwards the request to
``service2`` and also mirrors the request to the cluster named, ``service2-mirror``.

.. code-block:: console

  $ curl --header "x-mirror-cluster: service2-mirror" localhost:10000/service/2
  Hello from behind Envoy (service 2)!

Step 5: View logs for the request mirrored by request header
************************************************************

The logs show that both the ``service2`` and ``service2-mirror`` services
got the request.

.. code-block:: console

   $ pwd
   envoy/examples/route-mirror

   $ docker-compose logs service2
   ...
   Host: localhost:10000
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/2 HTTP/1.1" 200 -

   $ docker-compose logs service2-mirror
   ...
   Host: localhost-shadow:10000
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/2 HTTP/1.1" 200 -

You can also see that for the request to the ``service2-mirror`` service, the
``Host`` header was modified by Envoy to have a ``-shadow`` suffix in the

> hostname.

Author: That's right. I will move it.

Step 6: Missing or invalid cluster name in request header
*********************************************************

> If you do not specify the ``x-mirror-cluster`` in the request to ``service2``,
> a request to ``service1`` will not automatically be mirrored to ``service2-mirror``.

Reviewer: ```suggestion
If you do not specify the ``x-mirror-cluster`` in the request to ``service2``,
or specify an unknown cluster, the request will not be mirrored but will be handled in the normal way.
```

.. code-block:: console

  $ curl localhost:10000/service/2
  Hello from behind Envoy (service 2)!

View the logs for the ``service2`` and ``service2-mirror`` services.

.. code-block:: console

   $ pwd
   envoy/examples/route-mirror

   $ docker-compose logs service2
   ...
   Host: localhost:10000
   192.168.80.6 - - [06/Oct/2022 03:56:22] "GET /service/2 HTTP/1.1" 200 -

   $ docker-compose logs service2-mirror
   # No new logs

Similarly, if we specify a value for the ``x-mirror-cluster`` header such that there
is no such cluster defined with that name, ``service2-mirror`` will not receive

> the request:

Reviewer: i think collapse these so that you do all of the requests in the first code block and then you can separate out checking the logs to a separate step
